### ISSUE TITLE
Debug functional t5 model PCC

### ISSUE THREAD

**Issue by Sudharsan-V** [tenstorrent-metal/tt-metal#6342]
Update the t5 demo to be compatible with the ttnn_functional_t5 model.

**Comment by vigneshkeerthivasanx** [comment]
Demo for ttnn functional t5 is available in `models/experimental/functional_t5/demo/demo.py`.
While using the optimized version we are getting results

Case 1:
```
Batch_size = 1
Max_tokens = 5
Sequence_length = 128

Input json
t5-small:  1D mcast for in0 or in1 is not implemented yet.
google/flan-t5-small:  1D mcast for in0 or in1 is not implemented yet.

Input SQUAD V2:
t5-small:  1D mcast for in0 or in1 is not implemented yet.
google/flan-t5-small: 1D mcast for in0 or in1 is not implemented yet.
```

Case 2:
```
Batch_size = 2
Max_tokens = 5
Sequence_length = 128

Input json
t5-small: ['', 'chychychychy']
google/flan-t5-small: ['a a ', 'Norman']

Input SQUAD V2:
t5-small:  ['a', 'Who']
google/flan-t5-small: ['Norm "N', 'Norman']
```

Case 3:
For batch_size >=8. We are facing hang issue and the demo is failed due to timeout

Case 4:
```
With increased Sequence_length >= 384.
t5-small:    Matmul grid size exceeds maximum device compute grid size!
google/flan-t5-small:  Out of memory error
```


In `tests/ttnn/integration_tests/t5/test_ttnn_optimized_functional_t5.py` , it is checked with torch_random which returns zero tensors in test file which gives PCC = 0.99. But while testing those files with real index values we are getting low PCC , 0.61<PCC<0.95.

**Comment by boris-drazic** [comment]
@eyonland and @arakhmati do you have any suggestions how to quickly debug this PCC drop?

**Comment by vigneshkeerthivasanx** [comment]
When the test `tests/ttnn/integration_tests/t5/test_ttnn_optimized_functional_t5.py::test_t5_stack_decoder`  for  with tensor filled with non-zero values for t5-small , PCC = 0.8462392744
PCC analysis have been done by comparing with torch and ttnn_optimized model layer by layer.
t5_stack_decoder will utilize the t5_block sub-module in which we see the drop in PCC. And PFB the PCC for each block

```
block 0	0.8826683499
block 1	0.845905146
block 2	0.8200363768
block 3	0.8157465857
block 4	0.8150941374
block 5	0.8188650125
```

t5_block uses t5_self_attention, t5_cross_attention and t5_layer_ff. We see drop in PCC in t5_self_attention.

```
Self_attention 0	0.8763924666
Self_attention 1	0.8334680769
Self_attention 2	0.8144108553
Self_attention 3	0.8050339821
Self_attention 4	0.807878639
Self_attention 5	0.8120008994
```

self_attention modules uses t5_layer_norm and t5_attention.
We see drop in PCC in t5_layer_norm and t5_attention. We are currently debugging here to check the cause of PCC drop.

**Comment by vigneshkeerthivasanx** [comment]
ttnn functional t5 demo is moved under `models/demos/functional_t5/demo`
corresponding commit : [Link](https://github.com/tenstorrent-metal/tt-metal/compare/vignesh/move_ttnn_t5_demo_to_demos_folder)

**Comment by boris-drazic** [comment]
You should move entire folder for functional t5 to demos.
So `tt-metal/models/experimental/functional_t5/` moves to become `tt-metal/models/demo/functional_t5/`

**Comment by vigneshkeerthivasanx** [comment]
Layer by Layer analysis of the t5 model is done by passing non zero index values and found the reduction in PCC happening in t5_attention sub-module. By replacing the ttnn t5_attention submodule with torch t5_attention sub-module there is an improvement in the PCC. PFB the PCC analysis

**test_t5_for_conditional_generation**

- With ttnn attention sub-module

```
    t5-small : PCC:  0.527770122481286
    google/flan-t5-small:  PCC: 0.7030808092502201
```


- With torch attention sub-module 

```
    t5-small : PCC = 0.7683277007485174
    google/flan-t5-small : PCC = 0.8964956104291364
```


**test_t5_stack_decoder**

- With ttnn attention sub-module

```
    t5-small : PCC = 0.8462392744387687
    google/flan-t5-small: PCC = 0.655620748687943
```

- With torch attention sub-module 

```
    t5-small : PCC > 0.99
    google/flan-t5-small: PCC > 0.99
```

**Comment by Sudharsan-V** [comment]
While debugging the t5 pcc issue, we found there is a small discrepancy in the model used in the test file.
In test file `tests/ttnn/integration_tests/t5/test_ttnn_optimized_functional_t5.py` ([link](https://github.com/tenstorrent-metal/tt-metal/blob/main/tests/ttnn/integration_tests/t5/test_ttnn_optimized_functional_t5.py)), two models are used while testing.

- Case 1: transformers.models.t5.modeling_t5.T5Stack  in def test_t5_stack_decoder
- Case 2: transformers.T5ForConditionalGeneration.from_pretrained(model_name) in test_t5_for_conditional_generation

The major difference between the above-mentioned 2 cases are:

- In case 1, the layer_norm weights are filled with ones. 
- In case 2, the weights are pre-trained weights.

To find the op/layers that causes the reduction in the pcc, the ops/layers from the `ttnn.transformer` is replaced with its corresponding pytorch/ttnn ops. PFB the results below.

**Sub-module** : t5_stack_decoder

1. **model used**: transformers.models.t5.modeling_t5.T5Stack (Case 1 mentioned above)

- Without changes/replacement of op/layer:

```
Optimized (t5-small): 0.8462392744387687
Optimized: (flan-t5-small): 0.655620748687943
```

- Replaced ttnn.transformer.split_query_key_value_and_split_heads with equivalent pytorch codes:

```
Optimized (t5-small): 0.8462392744387687
Optimized: (flan-t5-small): 0.655620748687943
```

- Replaced ttnn.transformer.split_query_key_value_and_split_heads with equivalent pytorch codes and ttnn.transformer.attention_softmax_ with ttnn.softmax
```
Optimized (t5-small):0.9196367811387063
Optimized: (flan-t5-small): 0.7422650180056339
```

2. **model used**: transformers.T5ForConditionalGeneration.from_pretrained(model_name) (Case 2 mentioned above)

- Without changes/replacement of op/layer:

```
Optimized (t5-small): 0.8383127605948173
Optimized: (flan-t5-small): 0.6448584341369074
```

- Replaced ttnn.transformer.split_query_key_value_and_split_heads with equivalent pytorch codes:

```
Optimized (t5-small): 0.8383127605948173
Optimized: (flan-t5-small): 0.6448584341369074
```

- Replaced ttnn.transformer.split_query_key_value_and_split_heads with equivalent pytorch codes and ttnn.transformer.attention_softmax_ with ttnn.softmax

```
Optimized (t5-small):0.8425197364744049
Optimized: (flan-t5-small): 0.7494617281068209
```

**Sub-module** : t5_for_conditional_generation

1. **model used**: transformers.T5ForConditionalGeneration.from_pretrained(model_name) (Case 2 mentioned above)

- Without changes/replacement of op/layer:
```
Optimized (t5-small): 0.527770122481286
Optimized: (flan-t5-small): 0.7030808092502201
```

- Replaced ttnn.transformer.split_query_key_value_and_split_heads with equivalent pytorch codes and ttnn.transformer.attention_softmax_ with ttnn.softmax
```
Optimized (t5-small):0.7532525590057603
Optimized: (flan-t5-small): 0.7745995723532803
```

**Comment by vigneshkeerthivasanx** [comment]
all close comparison of output after each OP is tested and commit is available. Except Linear ops in t5_attention submodule have been tested as the weights are fused in ttnn_t5_optimized implementation.

corresponding commit  : [Link](https://github.com/tenstorrent-metal/tt-metal/commit/dd5e0dc85a6e662fc2744b62baeb854cb3daa908)

**Comment by boris-drazic** [comment]
@vigneshkeerthivasanx can you add test results for all close here?

**Comment by boris-drazic** [comment]
The values for all close in `t5_Attention` look concerning. RTOL for mat muls is very large and ATOL for add is huge.
Can you take a look at actual tensor values and what is the magnitude of them? Do these large differences make sense?
Is there only a few values that are very different and the rest are a good match?
For each of these three OPs, can you add to spreadsheet input and output tensors for each and matching tensors from PyTorch so we can plot and compare them value by value.
(It is probably enough to just put values for one selection of last two dims, i.e., [1, 1, :, :]).

**Comment by vigneshkeerthivasanx** [comment]
The observations, plots are recorded in the sheet - [Link](https://docs.google.com/document/d/10Yeqprgrxlwdj228q0QX6s9mvaa_EvlqxtQmR5jKPKc/edit)

**Comment by vigneshkeerthivasanx** [comment]
Commit has been updated with the unit test to to recreate the t5 plots that is recorded in the sheets.
corresponding commit : [Link](https://github.com/tenstorrent-metal/tt-metal/commit/aa0b84c7a28ee0ea68854e726b9ddcfa8261e69a)

**Comment by vigneshkeerthivasanx** [comment]
Demo file for t5 conditional generation using optimized ttnn t5 model is created.
Corresponding commit : [Link](https://github.com/tenstorrent-metal/tt-metal/compare/vignesh/demo_for_ttnn_t5_for_conditional_generation)

performance numbers for ttnn_optimized_t5 is checked and added in the sheets.
Sheets : [Link](https://docs.google.com/document/d/1DaY8ePkiA49NFhnkZ5h1qwaXT5TqP4es5QUJUvlGApo/edit)

**Comment by Sudharsan-V** [comment]
The pcc of the functional t5 model is improved.
The PR https://github.com/tenstorrent-metal/tt-metal/pull/7410 is merged with main.
CI tests passed [link](https://github.com/tenstorrent-metal/tt-metal/actions/runs/8663054761)
Closing this ticket
